pkg/vere/disk.c

```diff
@@ -867,66 +868,514 @@ u3_disk_init(c3_c* pax_c, u3_disk_cb cb_u)
+    //  migrate to the correct disk format if necessary
+    if ( c3n == u3_disk_migrate(log_u) ) {
+      fprintf(stderr, "disk: failed to migrate to v%d\r\n", U3D_VER1);
+      c3_free(log_u);
+      return 0;
     }
 
-    c3_free(log_c);
-  }
+    //  get latest epoch number
+    c3_d lat_d;
+    if ( c3n == u3_disk_epoc_last(log_u, &lat_d) ) {
+      fprintf(stderr, "disk: failed to load epoch number\r\n");
+      c3_free(log_u);
+      return 0;
+    }
 
-  //  get the latest event number from the db
-  //
-  {
-    log_u->dun_d = 0;
-    c3_d fir_d;
+    //  get binary version from latest epoch
+    c3_c ver_c[8193];
+    if ( c3n == u3_disk_epoc_vere(log_u, lat_d, ver_c) ) {
+      fprintf(stderr, "disk: failed to load epoch version\r\n");
+      c3_free(log_u);
+      return 0;
+    }
+
+    //  set path to latest epoch
+    c3_c epo_c[8193];
+    snprintf(epo_c, 8192, "%s/0i%" PRIc3_d, log_c, lat_d);
 
-    if ( c3n == u3_lmdb_gulf(log_u->mdb_u, &fir_d, &log_u->dun_d) ) {
-      fprintf(stderr, "disk: failed to load latest event from database\r\n");
+    //  initialize latest epoch's db
+    if ( 0 == (log_u->mdb_u = u3_lmdb_init(epo_c)) ) {
+      fprintf(stderr, "disk: failed to initialize database\r\n");
       c3_free(log_u);
       return 0;
     }
+    fprintf(stderr, "disk: loaded epoch 0i%" PRIc3_d "\r\n", lat_d);
 
+    //  get first/last event numbers from lmdb
+    c3_d fir_d, las_d;
+    if ( c3n == u3_lmdb_gulf(log_u->mdb_u, &fir_d, &las_d) ) {
+      fprintf(stderr, "disk: failed to get first/last event numbers\r\n");
+      return 0;
+    }
+
+    //  initialize dun_d/sen_d values
+    if ( 0 == las_d ) {               //  fresh epoch (no events in lmdb yet)
+      if ( 0 == lat_d ) {             //  first epoch
+        log_u->dun_d = 0;
+      } else {                        //  not first epoch
+        log_u->dun_d = lat_d;         //  set dun_d to last event in prev epoch
+      }
+    } else {                          //  not fresh epoch
+      log_u->dun_d = las_d;           //  set dun_d to last event in lmdb
+    }
     log_u->sen_d = log_u->dun_d;
+
+    //  if binary version of latest epoch is not the same as the
+    //  running binary, then we need to create a new epoch
+    if ( 0 != strcmp(ver_c, URBIT_VERSION) ) {
+      if ( c3n == u3_disk_epoc_init(log_u) ) {
+        fprintf(stderr, "disk: failed to initialize epoch\r\n");
+        c3_free(log_u);
+        return 0;
+      }
+    }
+
+    //  mark the log as live
+    log_u->liv_o = c3y;
   }
 
-  log_u->liv_o = c3y;
 
 #if defined(DISK_TRACE_JAM) || defined(DISK_TRACE_CUE)
   u3t_trace_open(pax_c);
 #endif
 
   return log_u;
 }
+
+/* u3_disk_epoc_good: check for valid epoch.
+*/
+c3_o
+u3_disk_epoc_good(u3_disk* log_u, c3_d epo_d) {
+  /*  a "valid" epoch is currently defined as a writable folder in
+   *  the <pier>/.urb/log directory named with a @ui and contains:
+   *  - a writable data.mdb file which can be opened by lmdb (add db checking?)
```

Comment:
example of the brace nit I mentioned above, though it looks like this applies to all new functions in disk.c